### ISSUE TITLE
NTFS junctions

### ISSUE THREAD

**Issue by jgoz** [coreybutler/nvm-windows#79]
Would it be feasible to add support for using NTFS junctions rather than `mklink` for swapping out the in-use version of node?

Junctions are great because they do not require elevated permissions, which would allow `nvm-windows` to be used on (effectively) headless machines, such as build agents. The disadvantage is that they cannot be created across drive letters, but this is an acceptable limitation.

Maybe this could be implemented as a configuration flag and disabled by default. Thoughts on that?

**Comment by jgoz** [comment]
Another point in favour of junctions: they are used by `npm` on Windows when using `npm link`, so it is already common in the Windows/node world.

**Comment by coreybutler** [comment]
I'll consider it for the next version (which I'm working on as time allows). I almost did this for the first releases, but I had a number of requests for cross-drive support, so I'm not sure if this is an acceptable limitation... but the idea of keeping it behind a flag is viable.

**Comment by SteveALee** [comment]
symbolic links are more flexible in the targets they support

**Comment by edclement** [comment]
> symbolic links are more flexible in the targets they support. see mklink.

some of are using workin machines though that restrict certain admin actions such as mklink

**Comment by SteveALee** [comment]
@edclement Not good for Dev PCs at all. Fight for freedom :)

**Comment by coreybutler** [comment]
@edclement - there is literally nothing that can be done if mklink isn't allowed by enterprise policy. I believe this would also affect npm. The entire premise of this project is based on the use of linked directories.

**Comment by edclement** [comment]
@coreybutler - I use `npm link` often. It actually uses junctions. It may be the case that it falls back to using windows junctions when it can't create a symlink. Not sure. I feel like nvm should do the same.

**Comment by jgoz** [comment]
@edclement, @coreybutler I'm reasonably sure that `npm link` uses junctions by default (which can be done with `mklink /J`, btw). Otherwise, linking would have to be done in an elevated console or `npm` would ship with an elevation script, similar to this project.

> there is literally nothing that can be done if mklink isn't allowed by enterprise policy

This is true, but I would be surprised if the `mklink` executable itself was disallowed. It's just the `/D` (create symbolic link) option that requires administrator privileges.

In my original issue description, I suggested toggling use of junction points with a flag, but I now think it should be the default option whenever the source & target drive letters are the same. But I'd be happy either way 😄

**Comment by coreybutler** [comment]
It does look like npm is using junctions for links, and it also looks like npm suffers from the same issue nvm4w would if it were installed on a separate partition.

I still like the idea of toggling, and I like @jgoz's follow up comment about checking the drive letters. I'm 99% sure I'll implement at least a toggle/flag if I can't find a better alternative to mklink.

**Comment by lukecummings** [comment]
modified elevate.cmd as a quick fix if you really need it
```
@setlocal
@echo off
set CMD=%*
set APP=%1

set newCMD=%CMD:mklink /d=mklink /j%
IF "%newCMD%" == "%CMD%" ( 
  start wscript //nologo "%~dpn0.vbs" %* & exit 
) ELSE ( 
  %newCMD%
)
```

**Comment by coreybutler** [comment]
I'm reposting this here since I inadvertantly merged it into the wiki some time ago. Written by Burt Harris:

## Elevator pitch for Junction Points

On Windows, a *[NTFS junction point](https://en.wikipedia.org/wiki/NTFS_junction_point)* can generally be created and deleted without any special privileges; junction points were created as a safer alternative to directory symlinks.   If **nvm-windows** switched its implementation to use NTFS junction points, it could would be simpler, more usable, and present lower security risks.   

## The demo

Using NTFS junctions is really simple, this should give you the idea:

``` 

C:\demo>dir
 Volume in drive C has no label.
 Volume Serial Number is 4C7F-6408

 Directory of C:\demo

07/09/2016  02:34 PM    <DIR>          .
07/09/2016  02:34 PM    <DIR>          ..
07/09/2016  02:31 PM    <DIR>          version.1
07/09/2016  02:28 PM    <DIR>          version.2
               0 File(s)              0 bytes
               4 Dir(s)  350,868,033,536 bytes free

C:\demo>mklink /j nodejs version.1
Junction created for nodejs <<===>> version.1

C:\demo>type nodejs\test.txt
This is version 1's test.txt

C:\demo>
C:\demo>rd nodejs

C:\demo>mklink /j nodejs version.2
Junction created for nodejs <<===>> version.2

C:\demo>type nodejs\test.txt
This is version 2's test.txt
```

As can be seen below, a junction point presents a much more understandable directory listing
```
C:\demo>dir
 Volume in drive C has no label.
 Volume Serial Number is 4C7F-6408

 Directory of C:\demo

07/09/2016  02:35 PM    <DIR>          .
07/09/2016  02:35 PM    <DIR>          ..
07/09/2016  02:35 PM    <JUNCTION>     nodejs [C:\demo\version.2]
07/09/2016  02:31 PM    <DIR>          version.1
07/09/2016  02:28 PM    <DIR>          version.2
               0 File(s)              0 bytes
               5 Dir(s)  350,868,033,536 bytes free

```

## Completing the picture

Ideally, one user on a machine installing or removing **nvm-windows** or Node.js versions using the tool shouldn't effect any other user's experience on the same machine.   For this reason, installing junctions in C:\Program Files would still require and administrator.   But per-user tools can generally be installed under the path in the APPDATA environment variable, and that's where the per-user nvm junction point should be located.   Similarly user-specific NVM_HOME, NVM_SYMLINK and PATH environment variables are preferable to machine-wide ones.

**Comment by veda-tschoening** [comment]
> The disadvantage is that they cannot be created across drive letters, but this is an acceptable limitation.

Are you sure about that? [MS docs and examples](https://learn.microsoft.com/en-us/windows/win32/fileio/hard-links-and-junctions#junctions) read different to me:

> A junction can also link directories located on different local volumes on the same computer.

> C:\dirx linked to D:\diry

The one not permitted example using `Z:` is because that is a mapped network drive and neither those nor UNC network paths are supported by junctions. So there's not really a practical restriction to use them at all.

**Comment by coreybutler** [comment]
@veda-tschoening Not being able to use a mapped network drive is a major problem for enterprise users who rely on a common SAN/NAS solution. Over 20% (~1.4M developers) of the user base would be negatively affected by this. That's why it has stalled.

NTFS junctions will not be coming to NVM4W. [Runtime](https://github.com/coreybutler/nvm-windows/wiki/Runtime), the successor we're working on, is planning to support a few different options.

**Comment by CloutKhan** [comment]
Would have been great if this could've been added years ago, stumbled across this facing the issue of requiring admin privileges to `use` making it unusable; I can still use this to manage the multiple versions, but have to manually create a junction link every time I would otherwise have used `use`. Although it seems silly that `mklink /D` always requires admin privileges to link a folder when it doesn't to link a file when having write permissions to the folder.

**Comment by mklemarczyk** [comment]
> @veda-tschoening Not being able to use a mapped network drive is a major problem for enterprise users who rely on a common SAN/NAS solution. Over 20% (~1.4M developers) of the user base would be negatively affected by this. That's why it has stalled.
> 
> NTFS junctions will not be coming to NVM4W. [Runtime](https://github.com/coreybutler/nvm-windows/wiki/Runtime), the successor we're working on, is planning to support a few different options though.

I suggest that is could be an option since that gives such limitations on both ends.
What @coreybutler do you think about that ? Other think is that program could detect the context in which we work. It is not that hard to check if the given location is a network drive in the end.

I'd like to work on the user local features for the NVM where the end user does not have much previliges, can not modify C:\Program files, and do not have rights to create symlinks that are not Junktions.

**Comment by mklemarczyk** [comment]
@jgoz I will create release with junctions on my fork, since owner do not want to maintain nvm-windows anymore.
https://github.com/mklemarczyk/nvm-windows

**Comment by coreybutler** [comment]
@mklemarczyk that's not really accurate. NVM for Windows will still be maintained until Runtime is released. There just won't be new features... that's what a feature freeze is. The new app is nearly ready for release.

**Comment by mklemarczyk** [comment]
@coreybutler keep working on new version. I hope it will improve dev life. Meantime I release the patched version on my fork repo. I hope it will help also others with the same constrain.

@jgoz  @edclement 
It is available to all under https://github.com/mklemarczyk/nvm-windows/releases/tag/v1.1.13

PS: If someone wants to add some new features, feel free to push PRs ;)

**Comment by mklemarczyk** [comment]
@coreybutler I see new major release has been done. Therefore it now accepting PRs ?

**Comment by mklemarczyk** [comment]
Yes it has been long time we hear about that mistic "Runtime" that has no preview, or publish appearance. Only a wiki page that is almost a year old.

**Comment by coreybutler** [comment]
@mklemarczyk I registered a legal entity to manage this last year.  I released 1.2.0 with part of Runtime in December. There have been some growing pains with this,  mostly around arm64, which I've been working on. I blogged about the transition to Runtime at https://opensource.author.io/nvm-for-windows-v120. There are thousands of people signed up for the newsletter, keeping up to date. I posted this in my summary reply above. Did you read any of that?
